controllers/check/check_iam.go

```diff
@@ -120,56 +118,107 @@ func getDeploymentStatus(r client.Reader, name string) string {
 	if err != nil {
 		if errors.IsNotFound(err) {
 			cm.Name = cmName
 			cm.Namespace = cmNs
 			cm.Data = make(map[string]string)
+			for _, requestNs := range requestNsSlice {
+				statusKey := requestNs + "-iamstatus"
+				cm.Data[statusKey] = status
+			}
 			cm.Data["iamstatus"] = status
-			if err := c.Create(context.TODO(), cm); err != nil {
+			if err := bs.Client.Create(context.TODO(), cm); err != nil {
 				klog.Errorf("Failed to create ConfigMap %s: %v", cmName, err)
 				return err
 			}
 			return nil
 		}
 		return err
 	}
-	if cm.Data["iamstatus"] != status {
-		klog.Infof("IAM status is %s", status)
+	isUpdate := false
+	for _, requestNs := range requestNsSlice {
+		statusKey := requestNs + "-iamstatus"
+		// check the iamstatus of cloud pak
+		if val, ok := cm.Data[statusKey]; ok {
+			if val != status {
+				klog.Infof("IAM status for namespace %s is %s", requestNs, status)
+				cm.Data[statusKey] = status
+				isUpdate = true
+			}
+		} else {
+			cm.Data[statusKey] = status
+			isUpdate = true
+		}
+	}
+
+	if isUpdate || cm.Data["iamstatus"] != status {
 		cm.Data["iamstatus"] = status
-		if err = c.Update(context.TODO(), cm); err != nil {
+		if status == "Ready" {
+			// iamstatus in this CS is ready, check iamstatus for all other CS
+			reg, _ := regexp.Compile(`^(.*)\-iamstatus`)
+			statusSlice := make([]string, 0)
+			for key, status := range cm.Data {
+				if reg.MatchString(key) {
+					statusSlice = append(statusSlice, status)
+				}
+			}
+			for _, status := range statusSlice {
+				if status != "Ready" {
+					cm.Data["iamstatus"] = status
+					break
+				}
+			}
+		}
+		if err = bs.Client.Update(context.TODO(), cm); err != nil {
 			klog.Errorf("Failed to update ConfigMap %s: %v", cmName, err)
 			return err
 		}
 	}
+
 	return nil
 }
 
-func updateConfigmap(r client.Reader, c client.Client, status string) error {
+func updateConfigmap(bs *bootstrap.Bootstrap, status string) error {
 	cm := &corev1.ConfigMap{}
 	cmName := "ibm-common-services-status"
 	cmNs := "kube-public"
-	err := r.Get(context.TODO(), types.NamespacedName{Name: cmName, Namespace: cmNs}, cm)
+	err := bs.Reader.Get(context.TODO(), types.NamespacedName{Name: cmName, Namespace: cmNs}, cm)
 	if errors.IsNotFound(err) {
 		return nil
 	} else if err != nil {
 		return err
 	}
-	if cm.Data["iamstatus"] != status {
-		klog.Infof("IAM status is %s", status)
-		cm.Data["iamstatus"] = status
-		if err = c.Update(context.TODO(), cm); err != nil {
+	requestNsSlice := util.GetRequestNs(bs.Reader)
+	isUpdate := false
+	for _, requestNs := range requestNsSlice {
+		// update the iam status for each cloud pak which is using this common service
+		statusKey := requestNs + "-iamstatus"
+		if val, ok := cm.Data[statusKey]; ok {
+			if val != status {
+				klog.Infof("IAM status for namespace %s is %s", requestNs, status)
+				cm.Data[statusKey] = status
+				cm.Data["iamstatus"] = status
```

Comment:
Should we update the `cm.Data["iamstatus"]` every time in a loop? That will cause it will always be the last status, isn't it?

Reply:
Oh is it because "status" is always not ready?

Could we abstract this part into another function and call the function here?

```
			// iamstatus in this CS is ready, check iamstatus for all other CS
			reg, _ := regexp.Compile(`^(.*)\-iamstatus`)
			statusSlice := make([]string, 0)
			for key, status := range cm.Data {
				if reg.MatchString(key) {
					statusSlice = append(statusSlice, status)
				}
			}
			for _, status := range statusSlice {
				if status != "Ready" {
					cm.Data["iamstatus"] = status
					break
				}
			}
```

Reply:
Yes, the function `updateConfigMap` will always get `NotReady` parameter, indicating that if cloud pak `iamstatus` is setting to `NotReady`, the overall `iamstatus` will be `NotReady` anyway.

But I think it will be better to generalize the logic of `updateConfigMap`, in order to make it follow the same logic with `CreateOrUpdateConfigMap`.

> Could we abstract this part into another function and call the function here?

Sure, it is updated, abstracting it into another function.

Please review the above changes again, thank you!